src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs

```diff
@@ -138,14 +138,18 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
             // order. We just need to iterate them and use the first one that can generate a link.
             var matches = _linkGenerationTree.GetMatches(context);
 
-            foreach (var match in matches)
+            // Perf: Avoid allocation of enumerator if matches is empty
+            if (matches?.Count > 0)
```

Comment:
Minor but the early exit you did [here](https://github.com/aspnet/Routing/pull/288/files#diff-653dd2a259cd4c89e150e818647183eaR42) is slightly cleaner. Nothing to block on however.

On the other hand, why is it necessary to exit early when `matches != null` regardless of the `Count`? The `for` loop isn't allocating an enumerator. Definitely shouldn't mention an enumerator in the `Perf` comment.